hardware/arduino/avr/cores/arduino/Print.h

```diff
@@ -57,28 +59,58 @@ class Print
     
     size_t print(const __FlashStringHelper *);
     size_t print(const String &);
-    size_t print(const char[]);
-    size_t print(char);
-    size_t print(unsigned char, int = DEC);
-    size_t print(int, int = DEC);
-    size_t print(unsigned int, int = DEC);
-    size_t print(long, int = DEC);
+    size_t print(  signed long, int = DEC);
     size_t print(unsigned long, int = DEC);
     size_t print(double, int = 2);
-    size_t print(const Printable&);
-
-    size_t println(const __FlashStringHelper *);
-    size_t println(const String &s);
-    size_t println(const char[]);
-    size_t println(char);
-    size_t println(unsigned char, int = DEC);
-    size_t println(int, int = DEC);
-    size_t println(unsigned int, int = DEC);
-    size_t println(long, int = DEC);
-    size_t println(unsigned long, int = DEC);
-    size_t println(double, int = 2);
-    size_t println(const Printable&);
+    
+    _always_inline size_t print(const char str[])   { return write(str); }
+    _always_inline size_t print(const char c)       { return write(c); }
+    _always_inline size_t print(const Printable &x) { return x.printTo(*this); }
+    
+    _always_inline size_t print(  signed char  n, int f = DEC) { return print((  signed long) n, f); }
+    _always_inline size_t print(  signed short n, int f = DEC) { return print((  signed long) n, f); }
+    _always_inline size_t print(  signed int   n, int f = DEC) { return print((  signed long) n, f); }
+    _always_inline size_t print(unsigned char  n, int f = DEC) { return print((unsigned long) n, f); }
+    _always_inline size_t print(unsigned short n, int f = DEC) { return print((unsigned long) n, f); }
+    _always_inline size_t print(unsigned int   n, int f = DEC) { return print((unsigned long) n, f); }
+    _always_inline size_t print(    float      n, int f = 2  ) { return print((    double   ) n, f); }
+    
     size_t println(void);
+    
+#if __cplusplus >= 201103L
+    template<typename ...Ts> _always_inline size_t println(const Ts &...args) { size_t t = print(args...); return t + println(); }
+#else
+    template<typename T> _always_inline size_t println(const T &arg)      { size_t t = print(arg);  return t + println(); }
+    template<typename T> _always_inline size_t println(const T &n, int f) { size_t t = print(n, f); return t + println(); }
+#endif // __cplusplus >= 201103L
+    
+    
+    /** Variadic methods **/
+#if __cplusplus >= 201103L  // requires C++11
+    // Ensure there are at least two parameters to avoid infinite recursion.
+    // e.g. `StringSumHelper s; print(s)` may be treated as `print(s, ...)` 
+    // with `...` being the empty list, thus calling `print(s)` again.
+    // (This is because print(StringSumHelper) isn't explicitly defined.)
```

Comment:
I don't agree with this - with one parameter, won't this simply fall back on `print()`, which you could define privately as a no-op?

Reply:
I recall that this was tried too, but that triggered other corner cases (that I can't recall right now). With care, it can be made to work though (I have some local changes on top of this PR that do so, as well as add some formatting support, but I can't get around to finishing them...).

Reply:
> with one parameter, won't this simply fall back on `print()`, which you could define privately as a no-op?

It will fall back on `print(s)` followed by `print()`.  The latter could indeed be implemented as a no-op, but the former will be expanded to `print(s)` followed by `print()` again, entering an infinite loop (at compile time).
I don't remember the exact situation that caused this (I think it was something about passing a `StringSumHelper` but I cannot come up with an example right now); only that infinite recursion happened.